[WIP]Support shared cache

LoserCheems opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #35876 


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

to: @zucchini-nlp 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @ylacombe, @eustlb
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

algo-home committed:
Support `SharedCache` (#35876)

algo-home committed:
Fix import of SharedCache

LoserCheems committed:
Merge branch 'huggingface:main' into support-shared-cache

LoserCheems committed:
Merge branch 'huggingface:main' into support-shared-cache

algo-home committed:
Add SharedCache to generation config

algo-home committed:
Optimized SharedCache

algo-home committed:
Add SharedCache tests

LoserCheems committed:
Fix code quality for `SharedCache`

zucchini-nlp commented:
Very interesting PR @LoserCheems , thanks for working on it! Before we review it, can you share any evaluations you did to verify that `SharedCache` is more memory-efficient and how much memory is saved, and how much it hurts model performance. I believe there's a sweet spot on `shared_layer_groups` which gives the best trade-off 

@gante

gante commented:
Hey @LoserCheems, thank you for opening the PR!

+1 to what @zucchini-nlp wrote -- some benchmarks are needed to back up the speedup and/or quality claims, along the obvious memory benefits 🤗

LoserCheems committed:
Fix init isort

LoserCheems commented:
Thank you @zucchini-nlp @gante, here are some evaluation tables in [this paper](https://arxiv.org/pdf/2411.13676).
![image](https://github.com/user-attachments/assets/9ae695a5-28bd-4407-9755-5e29d9f914ad)

zucchini-nlp commented:
@LoserCheems seems like the results you show are from a Hybrid Mamba-Transformer model. Could you write a small eval benchmark using key LLMs (e.g. llama, qwen) and test how the method applies to them

I’m a bit skeptical about how well the results from the paper transfer across different models while maintaining the same performance quality

LoserCheems commented:
Thank your suggest @zucchini-nlp, in the current test, I used `Llama-2-7b-hf` to conduct a comparison of memory usage with dynamic cache. However, I'm not quite sure how to write a quality benchmark test in `transformers`. I only know how to use `lighteval`.

zucchini-nlp commented:
You can take inspiration from a script we used for Static Cache (https://gist.github.com/ArthurZucker/ae0a86ef8f841c0ef69aaa52ccbc0b03), and remove the `torch.compile` part.

In general there are no guidelines, as long as the script compares the memory consumption of the default generation vs generation with proposed cache, using different `shared groups`, `batch sizes` and `max lengths`. For quality test, a simple perplexity measure would be enough on a dataset of your choice, to verify that Shared Cache doesn't degrade too much compared to baseline. Feel free to use the [`1B` llama](https://huggingface.co/meta-llama/Llama-3.2-1B-Instruct), if you don
t have much compute 🤗

LoserCheems commented:
Thank you @zucchini-nlp, but the benchmark submission will be late, because I currently only have a laptop with only CPU, and I also need to apply for the permission to use llama.

gante commented:
@LoserCheems a model without gating can also be used, e.g. qwen 2.5 :)

In a nutshell, we need to validate any technique before considering adding it to `transformers` -- otherwise, the library quickly gets bloated, which is bad for everyone. `SharedCache` seems indeed promising, but its original paper trained a model with this cache, i.e. we don't know whether it translates well to existing pre-trained models. If not, then we don't want it until we add a model with this cache.

A good benchmark would confirm that `SharedCache` doesn't lose a big amount of performance. Modifying `transformers` and evaluating the resulting code (e.g. with `lighteval`) is the strategy I recommend.

LoserCheems committed:
Merge branch 'huggingface:main' into support-shared-cache

algo-home committed:
Update last layer cache to be not shared
The cache of the last layer should not be shared, and it is the most different along with the first layer

algo-home committed:
Fix config error

algo-home committed:
Add

algo-home committed:
remove shared cache in NEED_SETUP_CACHE_CLASSES_MAPPING

LoserCheems committed:
Merge branch 'huggingface:main' into support-shared-cache
